### PR TITLE
mgr/nfs: add cmount_path

### PR DESCRIPTION
**Pull request by avanthakkar**
This PR implements the necessary changes for the NFS module to align with the new export FSAL block format introduced in `nfs-ganesha-V5.6`. The purpose of these changes is to enhance memory efficiency for exports. To achieve this goal, we have introduced a new field called `cmount_path` under the FSAL block of export. Initially, this is applicable only to CephFS-based exports.

Furthermore, the newly created CephFS exports will now share the same `user_id` and `secret_access_key`, which are determined based on the NFS cluster name and filesystem name. This results in each export on the same filesystem using a shared connection, thereby optimizing resource usage.
 

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
So far the changes allows the capability to create export with `cmount_path` and skips using the user_id in info
```
**$ ceph nfs export create cephfs nfs1 /cephfs abe**
{
  "bind": "/cephfs",
  "cluster": "nfs1",
  "fs": "abe",
  "mode": "RW",
  "path": "/"
}
**[ceph: root@ceph-node-00 /]# ceph nfs export info nfs1 /cephfs**
{
  "access_type": "RW",
  "clients": [],
  "cluster_id": "nfs1",
  "export_id": 1,
  "fsal": {
    "cmount_path": "/",
    "fs_name": "abe",
    "name": "CEPH"
  },
  "path": "/",
  "protocols": [
    4
  ],
  "pseudo": "/cephfs",
  "security_label": true,
  "squash": "none",
  "transports": [
    "TCP"
  ]
}
```

File: src/pybind/mgr/nfs/ganesha_conf.py [hunk 1]
```diff
@@ -208,50 +209,52 @@ def to_dict(self) -> Dict[str, Any]:
 class CephFSFSAL(FSAL):
     def __init__(self,
                  name: str,
-                 user_id: Optional[str] = None,
+                 cmount_path: Optional[str] = None,
+                #  user_id: Optional[str] = None,
                  fs_name: Optional[str] = None,
-                 sec_label_xattr: Optional[str] = None,
-                 cephx_key: Optional[str] = None) -> None:
```
Comment (phlogistonjohn) [inline src/pybind/mgr/nfs/ganesha_conf.py:214]: As discussed in downstream slack, I think its OK to start with a branch of quick-and-dirty changes, however I think the core types like FSAL/CephFSFSAL can probably continue being fully compatible with all kinds of ganesha configs. Since user_id and cephx_key can already be None, I don't think it is necessary to comment it out.

Also, we can add cmount_path to the end of the args list.

File: src/pybind/mgr/nfs/ganesha_conf.py [hunk 2]
```diff
@@ -208,50 +209,52 @@ def to_dict(self) -> Dict[str, Any]:
...
+                 sec_label_xattr: Optional[str] = None):
         super().__init__(name)
         assert name == 'CEPH'
+        self.cmount_path = cmount_path if cmount_path else "/"
```
Comment (phlogistonjohn) [inline src/pybind/mgr/nfs/ganesha_conf.py:218]: Since cmount_path is a string we can simply default to it in the args like `cmount_path: Optional[str] = "/"`. Then explicitly passing a None or empty string is a way the caller can indicate that cmount_path could *not* be used for this item.

File: src/pybind/mgr/nfs/ganesha_conf.py [hunk 3]
```diff
@@ -295,12 +301,13 @@ def to_fsal_block(self) -> RawBlock:
...
     def from_dict(cls, fsal_dict: Dict[str, str]) -> 'RGWFSAL':
         return cls(fsal_dict['name'],
+                   fsal_dict['cmount_path'],
                    fsal_dict.get('user_id'),
                    fsal_dict.get('access_key_id'),
                    fsal_dict.get('secret_access_key'))
 
     def to_dict(self) -> Dict[str, str]:
-        r = {'name': self.name}
+        r = {'name': self.name, 'cmount_path': self.cmount_path}
```
Comment (phlogistonjohn) [inline src/pybind/mgr/nfs/ganesha_conf.py:310]:
```suggestion
        r = {'name': self.name}
        if self.cmount_path:
            r['cmount_path'] = self.cmount_path
```

Comment (avanthakkar) [reply src/pybind/mgr/nfs/ganesha_conf.py]: Okay! But anyways default value we're setting as `/` so it'll never be `None` right?

Comment (phlogistonjohn) [reply src/pybind/mgr/nfs/ganesha_conf.py]: If we read and parsed an old object from a previous build/version of ceph it may not have a cmount_path. I'd like to not break upstream users that might have deployed nfs in reef for example, unless absolutely necessary.

Comment (avanthakkar) [reply src/pybind/mgr/nfs/ganesha_conf.py]: Ah! I see the point here, completely missed! Thanks for clarification @phlogistonjohn. I'll update the PR

File: src/script/cpatch.py [hunk 4]
```diff
@@ -527,6 +527,7 @@ def _py_site_packages(self):
             return self._cached_py_site_packages
         # use the container image to probe for the correct python site-packages dir
         valid_site_packages = [
+            "/usr/lib/python3.9/site-packages",
```
Comment (phlogistonjohn) [inline src/script/cpatch.py:530]: Put this in a separate commit please.

File: src/pybind/mgr/nfs/ganesha_conf.py [hunk 5]
```diff
@@ -243,15 +249,15 @@ def to_fsal_block(self) -> RawBlock:
     @classmethod
     def from_dict(cls, fsal_dict: Dict[str, Any]) -> 'CephFSFSAL':
         return cls(fsal_dict['name'],
+                   fsal_dict['cmount_path'],
```
Comment (phlogistonjohn) [inline src/pybind/mgr/nfs/ganesha_conf.py:252]: I don't expect this function to work correctly. The 2nd argument to the __init__ function is (still) user_id. I suggest changing this `from_dict` call to use keyword arguments and avoid having to reshuffle argument order.

File: src/pybind/mgr/nfs/export.py [hunk 6]
```diff
@@ -784,7 +780,8 @@ def _apply_export(
         )
 
         if not old_export:
-            self._create_export_user(new_export)
+            if old_export.fsal.name == NFS_GANESHA_SUPPORTED_FSALS[1]:
+                self._create_export_user(new_export)
```
Comment (phlogistonjohn) [inline src/pybind/mgr/nfs/export.py:784]: I'm not sure what this call to `_create_export_user` is doing. Can you explain or add a comment?

File: src/pybind/mgr/cephadm/module.py [hunk 7]
```diff
@@ -104,7 +104,7 @@ def os_exit_noop(status: int) -> None:
 
 
 # Default container images -----------------------------------------------------
-DEFAULT_IMAGE = 'quay.io/ceph/ceph'
+DEFAULT_IMAGE = 'quay.io/athakkar/nfs'
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/module.py:107]: Will need to be removed before merge or run thru teuthology.

**Review by phlogistonjohn — COMMENTED** [review]
I put a few comments. I think for next steps we need to write or update some of the unit tests to make sure the export create/apply/remove operations handle both old and new style exports.

File: src/pybind/mgr/nfs/export.py (hunk 6 above)
Comment (avanthakkar) [reply src/pybind/mgr/nfs/export.py:784]: We're creating new user(with `user_id`) for every export & corresponding cephx key. Since we no more this for CephFS based nfs exports I put condition only for RGW ones.

File: src/pybind/mgr/cephadm/module.py (hunk 7 above)
Comment (avanthakkar) [reply src/pybind/mgr/cephadm/module.py:107]: Yes! I'll remove these changes.

File: src/pybind/mgr/nfs/export.py (hunk 6 above)
Comment (phlogistonjohn) [reply src/pybind/mgr/nfs/export.py:784]: Ah, got it. The `NFS_GANESHA_SUPPORTED_FSALS[1]` obfuscates the meaning the line. You might want to add a commend like `# only for rgw` but I won't demand it. Maybe some day we can come back in and use clearer constants like `FSAL_RGW` instead of `NFS_GANESHA_SUPPORTED_FSALS[1]`

Comment (avanthakkar) [reply src/pybind/mgr/nfs/export.py:784]: Sure, I'll add the comment atleast.

File: src/pybind/mgr/nfs/ganesha_conf.py (hunk 5 above)
Comment (phlogistonjohn) [reply src/pybind/mgr/nfs/ganesha_conf.py:252]: I tried testing the current patches. I hit the following error. I suspect it may be related to this change:
```
[nfs ERROR nfs.export] Failed to create /foo export for nfs1
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/nfs/export.py", line 401, in create_export
    return self.create_cephfs_export(**kwargs)
  File "/usr/share/ceph/mgr/nfs/export.py", line 671, in create_cephfs_export
    export = self.create_export_from_dict(
  File "/usr/share/ceph/mgr/nfs/export.py", line 642, in create_export_from_dict
    export.validate(self.mgr)
  File "/usr/share/ceph/mgr/nfs/ganesha_conf.py", line 511, in validate
    raise FSNotFound(fs.fs_name)
nfs.exception.FSNotFound: filesystem None not found
```

**Comment by phlogistonjohn** [comment]
Hi @avanthakkar I worked more on this and instead of commenting I needed to get more hands on. I have changes in my branch here: https://github.com/phlogistonjohn/ceph/commits/jjm-nfs-export-orch-changes

I tried to push them into your branch but I got an access denied error. Feel free to cherry-pick from my branch as needed. We can also discuss in more detail later.

**Comment by avanthakkar** [comment]
> Hi @avanthakkar I worked more on this and instead of commenting I needed to get more hands on. I have changes in my branch here: https://github.com/phlogistonjohn/ceph/commits/jjm-nfs-export-orch-changes
> 
> I tried to push them into your branch but I got an access denied error. Feel free to cherry-pick from my branch as needed. We can also discuss in more detail later.

It's probably because you may not be authorised to commit to[ rhcs-dashboard/ceph](https://github.com/rhcs-dashboard/ceph) repo where `ceph-dashboard` team pushes all  their branches.  
Yeah I'll try to cherry-pick your commits and test it locally, thanks!

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by adk3798 — APPROVED** [review]
LGTM given my understanding of the requirements here (having cmount_path in the FSAL and only creating one user per FS in NFS over cephFS case)

**Comment by BlaineEXE** [comment]
Could someone explain more what `cmount_path` is and means? This will Ceph-CSI and might affect Rook, and I would like to understand how to move forward with integrating these changes into those projects. Is there a discussion elsewhere someone can direct me to?

**Comment by avanthakkar** [comment]
Created the Ceph tracker  for this & updated the commit message

**Comment by avanthakkar** [comment]
> Could someone explain more what `cmount_path` is and means? This will Ceph-CSI and might affect Rook, and I would like to understand how to move forward with integrating these changes into those projects. Is there a discussion elsewhere someone can direct me to?

I think this is the initial work started on nfs-ganesha to introduce these changes https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/1170100.

**Comment by adk3798** [comment]
> I ran this through a partial teuth run of just nfs related tests. That run would have had these changes but still an older version of nfs-ganesha (specifically 5.5 I think). It actually only introduced a single extra test failure, so I guess this patch works okay with a ganesha version without all the cmount stuff.
> 
> The only new failure was https://pulpito.ceph.com/adking-2023-11-03_15:06:55-orch:cephadm-wip-adk4-testing-2023-11-01-1136-distro-default-smithi/7445588 where it failed with
> 
> ```
> 2023-11-03T15:32:32.832 INFO:tasks.cephfs_test_runner:Test successful creation and deletion of the cephfs export. ... FAIL
> 2023-11-03T15:32:32.833 INFO:tasks.cephfs_test_runner:
> 2023-11-03T15:32:32.833 INFO:tasks.cephfs_test_runner:======================================================================
> 2023-11-03T15:32:32.834 INFO:tasks.cephfs_test_runner:FAIL: test_create_and_delete_export (tasks.cephfs.test_nfs.TestNFS)
> 2023-11-03T15:32:32.834 INFO:tasks.cephfs_test_runner:Test successful creation and deletion of the cephfs export.
> 2023-11-03T15:32:32.834 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2023-11-03T15:32:32.835 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2023-11-03T15:32:32.835 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_24d91223074e2d4433876d8d08f0e98967545ab6/qa/tasks/cephfs/test_nfs.py", line 441, in test_create_and_delete_export
> 2023-11-03T15:32:32.835 INFO:tasks.cephfs_test_runner:    self._create_default_export()
> 2023-11-03T15:32:32.835 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_24d91223074e2d4433876d8d08f0e98967545ab6/qa/tasks/cephfs/test_nfs.py", line 231, in _create_default_export
> 2023-11-03T15:32:32.836 INFO:tasks.cephfs_test_runner:    self._create_export(export_id='1', create_fs=True)
> 2023-11-03T15:32:32.836 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_24d91223074e2d4433876d8d08f0e98967545ab6/qa/tasks/cephfs/test_nfs.py", line 219, in _create_export
> 2023-11-03T15:32:32.836 INFO:tasks.cephfs_test_runner:    self._check_auth_ls(export_id, check_in=True)
> 2023-11-03T15:32:32.837 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_24d91223074e2d4433876d8d08f0e98967545ab6/qa/tasks/cephfs/test_nfs.py", line 130, in _check_auth_ls
> 2023-11-03T15:32:32.837 INFO:tasks.cephfs_test_runner:    self.assertIn(f'{client_id}.{export_id}', output)
> 2023-11-03T15:32:32.837 INFO:tasks.cephfs_test_runner:AssertionError: 'client.nfs.test.1' not found in 'mds.a.smithi019.kehujq\n\tkey: AQClEUVl7nitIxAAEUAQkDsdFKfjSqjk5+Qo3g==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nmds.a.smithi019.niympu\n\tkey: AQCjEUVleNM4NxAAZFM75dOpONXVW6/LBK8hrw==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nmds.nfs-cephfs.smithi019.dgsvpn\n\tkey: AQCJEkVlnfGBLhAALH8MzUazJYnwKBH4lUJlAQ==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nmds.nfs-cephfs.smithi019.wsxuyz\n\tkey: AQCLEkVlA77wHRAARCABZFNEjAUZTXkZ9lS1rw==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nosd.0\n\tkey: AQBeEUVlKvQWLRAAXoWSh07c66WUseeLeCfB8w==\n\tcaps: [mgr] allow profile osd\n\tcaps: [mon] allow profile osd\n\tcaps: [osd] allow *\nosd.1\n\tkey: AQBuEUVlPgvNIBAAwZLMQ2w1doHDPNUkNz+jww==\n\tcaps: [mgr] allow profile osd\n\tcaps: [mon] allow profile osd\n\tcaps: [osd] allow *\nosd.2\n\tkey: AQB+EUVlV1eyGBAAO5OCizQlgv8QLj79uNtIMg==\n\tcaps: [mgr] allow profile osd\n\tcaps: [mon] allow profile osd\n\tcaps: [osd] allow *\nclient.0\n\tkey: AQCSEUVlcbRgIBAAJ5ymLuW41kcg8w0/cUuWWg==\n\tcaps: [mds] allow *\n\tcaps: [mgr] allow *\n\tcaps: [mon] allow *\n\tcaps: [osd] allow *\nclient.admin\n\tkey: AQAVEUVlNJXmHRAAFbQGTgSkhs0oKoAsSlqzPQ==\n\tcaps: [mds] allow *\n\tcaps: [mgr] allow *\n\tcaps: [mon] allow *\n\tcaps: [osd] allow *\nclient.bootstrap-mds\n\tkey: AQAZEUVlHXUPBBAAsbEDISnTmv1Fhl9P3iYRpw==\n\tcaps: [mon] allow profile bootstrap-mds\nclient.bootstrap-mgr\n\tkey: AQAZEUVly30PBBAAzyU0qpDH17xi3j2+8Xo+IQ==\n\tcaps: [mon] allow profile bootstrap-mgr\nclient.bootstrap-osd\n\tkey: AQAZEUVlk4YPBBAAEr1xivmRWzk0O3PLi47N/w==\n\tcaps: [mon] allow profile bootstrap-osd\nclient.bootstrap-rbd\n\tkey: AQAZEUVlX44PBBAAkWE4Qw0q2X3D9X3f6cLGvA==\n\tcaps: [mon] allow profile bootstrap-rbd\nclient.bootstrap-rbd-mirror\n\tkey: AQAZEUVlPpYPBBAAkwcPr7iys5gSfMyDKHknew==\n\tcaps: [mon] allow profile bootstrap-rbd-mirror\nclient.bootstrap-rgw\n\tkey: AQAZEUVla6APBBAAx0sjdMuuA/esk6E/Ou97yw==\n\tcaps: [mon] allow profile bootstrap-rgw\nclient.nfs.test.0.0.smithi019.dzbsco\n\tkey: AQB/EkVl5Z4ILRAAWAwScd3gprUDip0lVdY4JA==\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rw pool=.nfs namespace=test\nclient.nfs.test.0.0.smithi019.dzbsco-rgw\n\tkey: AQB/EkVl88ZjMxAA87734gTvS6Blo2Wg3r7WSA==\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rwx tag rgw *=*\nclient.nfs.test.nfs-cephfs\n\tkey: AQCPEkVlpl1YEhAAEvoPolQP/kEudM9xFP+qBg==\n\tcaps: [mds] allow rw path=/\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rw pool=.nfs namespace=test, allow rw tag cephfs data=nfs-cephfs\nclient.test\n\tkey: AQAHEkVlYj/JJBAArSxmGGnjWObYkxm8L67cWw==\n\tcaps: [mds] allow rw path=/\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rw pool=.nfs namespace=test, allow rw tag cephfs data=user_test_fs\nmgr.x\n\tkey: AQAWEUVl0fMeIhAAWfCR2Y+GP2ZTUWOfu3UEeQ==\n\tcaps: [mds] allow *\n\tcaps: [mon] profile mgr\n\tcaps: [osd] allow *\n'
> ```

the full run of nfs tests https://pulpito.ceph.com/adking-2023-11-03_13:20:43-orch:cephadm-wip-adk4-testing-2023-11-01-1136-distro-default-smithi/
and reruns (since 2 of the failures were https://tracker.ceph.com/issues/49287 which just happens occasionally) https://pulpito.ceph.com/adking-2023-11-03_15:06:55-orch:cephadm-wip-adk4-testing-2023-11-01-1136-distro-default-smithi/

**Comment by phlogistonjohn** [comment]
`AssertionError: 'client.nfs.test.1' not found in` seems to indicate that the test is looking specifically for the old style of cephx user where "test" is the name of the cluster and "1" is the export id. This means that test needs updating because the user scheme has changed.

**Comment by nixpanic** [comment]
Is there a need to migrate previous NFS-Ganesha configurations to a new format? Or does the old format keep working, even after updating the Ceph container-images with a new NFS-Ganesha version?

**Comment by avanthakkar** [comment]
> Is there a need to migrate previous NFS-Ganesha configurations to a new format? Or does the old format keep working, even after updating the Ceph container-images with a new NFS-Ganesha version?

No need. We have backward compatibility for older format & works as before w/o any configuration changes.
Correct me if I'm wrong @phlogistonjohn

**Comment by phlogistonjohn** [comment]
There's two parts to the backwards compatibility: the mgr module needs to be able to read "old style" configs from the rados pool. We think we've covered that via unit tests in the mgr module. The other is that ganehsa needs to also be able to read those older configs from the mgr module - we've been told that's the case but I think we should try to verify that even if it's done as a one off task and not part of a test suite. (I've heard rumors of some builds not working correctly)

**Comment by nixpanic** [comment]
There is a suspicion that a newer version of NFS-Ganesha does not work with current (old) configuration snippets that are created by Ceph Mgr. Failures with NFS-Ganesha 5.7 are included in https://github.com/rook/rook/issues/13151#issuecomment-1795192389.

**Comment by avanthakkar** [comment]
Closing in favor of https://github.com/ceph/ceph/pull/54277#
